### ISSUE TITLE
Memory fragmentation in threaded executors

### ISSUE THREAD

**Issue by nsmith-** [scikit-hep/coffea#249]
I ran into a really interesting “leak” while diagnosing high memory usage in threaded execution environments with coffea.  Basically the glibc memory allocation system has some dynamic threshold at which it decides whether to do an actual syscall to `mmap` vs when it uses a pre-existing arena (or creates a new larger arena) to allocate memory for a new object (`malloc`) When using many large numpy arrays and constantly constructing and destructing them, one quickly teaches glibc malloc to prefer very large arenas to prevent expensive mmap syscalls.  This is all fine in single threaded mode because the large numpy temporory array mallocs and frees are close to each other and not much other long-lasting stuff will get a chance to malloc in between them.  In a threaded execution environment, another thread might call malloc to allocate a long-lived object in between a temporary numpy allocation, pinning the glibc arena from being released back to the OS even if the arena is mostly empty.  This leads to an apparent memory “leak” as the real RSS as reported by the system refuses to go down after all tasks are done and all the large numpy objects are cleaned up, which is really just tons of almost-empty arenas that glibc is not smart enough to patch up.  Some links that I found while digging this include [this thread on dask](https://github.com/dask/dask/issues/3530) and this [python bug report](https://bugs.python.org/issue33444) and a small [test example](https://gist.github.com/pitrou/6c5310d4c721436165666044e3c31158) demonstrating the effect therein. I copied the nice glibc mallinfo summary function from [here](https://github.com/dask/dask/issues/3530#issuecomment-393576996) into my context and indeed at the end of the program the memory fragmentation was severe:
```
Current RSS: 1.73 GB
Malloc info:
Total non-mmapped bytes (arena):       1.60 GB
# of free chunks (ordblks):            17.87 kB
# of free fastbin blocks (smblks):     92.11 kB
# of mapped regions (hblks):           1 B
Bytes in mapped regions (hblkhd):      2.63 MB
Max. total allocated space (usmblks):  0 B
Free bytes held in fastbins (fsmblks): 5.07 MB
Total allocated space (uordblks):      24.76 MB
Total free space (fordblks):           1.58 GB
Topmost releasable block (keepcost):   8.94 MB
```
showing that there is 1.58 GB tied up.  Suggestions in the thread that help alleviate the pressure include forcing glibc to use a static threshold for when to use `mmap` syscall, using the `MALLOC_MMAP_THRESHOLD_` environment variable, set to something smaller than the typical numpy array size, and also to try an alternative allocator such as jemalloc.  The former worked well, and I haven’t had a chance to try the latter yet.

Perhaps in a production cluster, it would be good to make sure the workers run with a better allocator, or take the performance hit of using a static mmap threshold

**Comment by nsmith-** [comment]
As more dask users are running into this issue, someone has kindly written up some additional discussion in https://github.com/dask/distributed/pull/4874 which can be found in https://distributed.dask.org/en/stable/worker-memory.html#memory-not-released-back-to-the-os

**Comment by yimuchen** [comment]
I tried all the non-jemalloc suggestions in the distributed doc: 

- `export MALLOC_TRIM_THRESHOLD_=1024`
- `export MALLOC_MMAP_THRESHOLD_=1024`
- Adding a manual malloc_trim call at the end of a `run_uproot_job` call. 

The memory footprint for all workers still creeps up indefinitely, with the majority of the memory being labelled as unmanaged.

**Comment by yimuchen** [comment]
I did some tests locally with the `iterative_executor`, and I did see a small memory creep over 100 chunks, but it is on the order of 10 MB, not the 1 GB creep we are seeing with `dask_executors`. I'm tentatively concluding that this has something to do with how dask workers are managing/tracking memory, not a direct issue with coffea/awkward.

**Comment by andrzejnovak** [comment]
@yimuchen are you using casa? If not you could try parsl instead of dask, I've found it to be more resilient

**Comment by yimuchen** [comment]
@andrzejnovak I'm mainly using lpcjobqueue on LPC, so unfortunately that isn't an option. 
Ultimately, I found that my hack of using [self-as-first-argument](https://stackoverflow.com/questions/972/adding-a-method-to-an-existing-object-instance/28060251#28060251) to split up the processor into multiple notebook cells was a prime cause of the memory issue. Refactored that part, the memory creep issue is still present, but much more manageable.

**Comment by yimuchen** [comment]
This is now known to be fixed with awkward==1.10.3
